### PR TITLE
Closing a channel should cleanup recorded auto-delete queues

### PR DESCRIPTION
**Pull request by vikinghawk**
## Proposed Changes

Closing a channel should remove recorded auto-delete queues just like it does for explicitly canceling a consumer.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

**Comment by michaelklishin** [comment]
CI reported a lot of failures like this:

```
04:35:02.605 [rabbitmq-connection-shutdown-amqp://guest@127.0.0.1:5672/] ERROR c.r.c.impl.ForgivingExceptionHandler - Caught an exception when recovering topology Caught an exception while recovering consumer amq.ctag-z7lKh0mMzJkcQtj0Uw5elg: channel is already closed due to channel error; protocol method: #method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no queue '-2745730218113755' in vhost '/', class-id=60, method-id=20)
com.rabbitmq.client.TopologyRecoveryException: Caught an exception while recovering consumer amq.ctag-z7lKh0mMzJkcQtj0Uw5elg: channel is already closed due to channel error; protocol method: #method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no queue '-2745730218113755' in vhost '/', class-id=60, method-id=20)
	at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.recoverConsumer(AutorecoveringConnection.java:821)
	at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.recoverTopology(AutorecoveringConnection.java:681)
	at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.beginAutomaticRecovery(AutorecoveringConnection.java:583)
	at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.lambda$addAutomaticRecoveryListener$3(AutorecoveringConnection.java:509)
	at com.rabbitmq.client.impl.AMQConnection.notifyRecoveryCanBeginListeners(AMQConnection.java:808)
	at com.rabbitmq.client.impl.AMQConnection.doFinalShutdown(AMQConnection.java:785)
	at com.rabbitmq.client.impl.nio.NioLoop.lambda$dispatchShutdownToConnection$2(NioLoop.java:305)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.rabbitmq.client.AlreadyClosedException: channel is already closed due to channel error; protocol method: #method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no queue '-2745730218113755' in vhost '/', class-id=60, method-id=20)
```

This can be due to our recent test suite changes as I'm running into an unrelated error in `master`.

**Comment by michaelklishin** [comment]
```
[ERROR]   ConnectionRecovery>BrokerTestCase.tearDown:94->BrokerTestCase.closeChannel:171 » ConcurrentModification
[ERROR]   ConnectionRecovery.consumersAreRemovedFromConnectionWhenChannelIsClosed:792 » ConcurrentModification
```

is an error specific to this branch.

**Comment by michaelklishin** [comment]
I used

```
./mvnw verify -P '!setup-test-cluster',use-nio -Dtest-broker.A.nodename=rabbit@$(hostname -s) -Drabbitmqctl.bin=rabbitmqctl -Dit.test=ConnectionRecovery
```

to run the suite.

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ private void executeAndClean(IoTimeoutExceptionRunnable callback) throws IOExcep
             callback.run();
         } finally {
             for (String consumerTag : consumerTags) {
-                this.connection.deleteRecordedConsumer(consumerTag);
+                this.deleteRecordedConsumer(consumerTag);
```
Comment (acogoluegnes) [inline src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java:104]: Iterating over the consumer tags set and removing them in `deleteRecordedConsumer(String)` causes a `ConcurrentModificationException`. Maybe moving the removal to the caller is a solution (and using an iterator and `Iterator#remove()` in the iteration). Not super-DRY, but it should work. The channel instance will be GC-ed in the case of the closing anyway.

**Review by acogoluegnes — CHANGES_REQUESTED** [review]

**Review by acogoluegnes — APPROVED** [review]

**Comment by acogoluegnes** [comment]
Thanks!

**Pull request merged**